src/System.Net.Security/src/System/Net/Security/SniHelper.cs

```diff
@@ -0,0 +1,278 @@
+        {
+            if (serverNameListExtension.Length < 2)
+            {
+                remainingBytes = ReadOnlySpan<byte>.Empty;
+                return null;
+            }
+
+            int serverNameListLength = ReadUint16(serverNameListExtension);
+            ReadOnlySpan<byte> serverNameList = serverNameListExtension.Slice(2);
+
+            if (serverNameListLength > serverNameList.Length)
+            {
+                remainingBytes = ReadOnlySpan<byte>.Empty;
+                return null;
+            }
+
+            remainingBytes = serverNameList.Slice(serverNameListLength);
+            ReadOnlySpan<byte> serverName = serverNameList.Slice(0, serverNameListLength);
+
+            if (serverName.Length < 3)
+            {
+                return null;
+            }
+
+            // -1 for hostNameType
+            int hostNameStructLength = ReadUint16(serverName) - 1;
+            byte hostNameType = serverName[2];
+            ReadOnlySpan<byte> hostNameStruct = serverName.Slice(3);
+
+            if (hostNameStructLength != hostNameStruct.Length || hostNameType != 0x00)
+            {
+                return null;
+            }
+
+            int hostNameLength = ReadUint16(hostNameStruct);
+            ReadOnlySpan<byte> hostName = hostNameStruct.Slice(2);
+
+            return DecodeString(hostName);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static string DecodeString(ReadOnlySpan<byte> bytes)
+        {
+            string idnEncodedString = Encoding.UTF8.GetString(bytes);
+            try
+            {
+                return s_idnMapping.GetUnicode(idnEncodedString);
+            }
+            catch (ArgumentException)
+            {
+                // client has not done IDN mapping
+                return idnEncodedString;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int ReadUint16(ReadOnlySpan<byte> bytes)
+        {
+            return (bytes[0] << 8) | bytes[1];
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int ReadUint24(ReadOnlySpan<byte> bytes)
+        {
+            return (bytes[0] << 16) | (bytes[1] << 8) | bytes[2];
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static ReadOnlySpan<byte> SkipBytes(ReadOnlySpan<byte> bytes, int numberOfBytesToSkip)
+        {
+            return (numberOfBytesToSkip < bytes.Length) ? bytes.Slice(numberOfBytesToSkip) : ReadOnlySpan<byte>.Empty;
+        }
+
+        // Opaque type is of structure:
+        //   - length (minimum number of bytes to hold the max value)
+        //   - data (length bytes)
+        // We will only use opaque types which are of max size: 255 (length = 1) or 2^16-1 (length = 2).
+        // We will call them SkipOpaqueType`length`
+        private static ReadOnlySpan<byte> SkipOpaqueType1(ReadOnlySpan<byte> bytes)
```

Comment:
On the one hand I wonder if it'd be better to have `SkipOpaque(ReadOnlySpan<byte> bytes, uint maxSize)` which does the log256 calculation.  But there's little enough here that it's fine, I guess.